File: UsbSerialConsole/app/src/main/java/jp/sugnakys/usbserialconsole/LogViewActivity.java
package jp.sugnakys.usbserialconsole;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.charset.Charset;

import jp.sugnakys.usbserialconsole.util.Constants;

public class LogViewActivity extends BaseAppCompatActivity {

    private static final String TAG = "LogViewActivity";

    private File logFile;
    private Toolbar toolbar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.log_view_main);
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            toolbar.setNavigationOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    finish();
                }
            });
        }
    }

    public void onResume() {
        super.onResume();

        logFile = (File) getIntent().getExtras().get(Constants.EXTRA_LOG_FILE);
        if (logFile != null) {
            toolbar.setTitle(logFile.getName());
            setLogText(logFile);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_log_view, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Intent intent;
        switch (item.getItemId()) {
            case R.id.action_send_to:
                intent = new Intent();
                intent.setAction(Intent.ACTION_VIEW);
                intent.setDataAndType(Uri.fromFile(logFile), "text/plain");
                startActivity(intent);
                break;
            default:
                Log.e(TAG, "Unknown id");
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setLogText(File file) {
        TextView textView = (TextView) findViewById(R.id.logView);
        FileInputStream fileInputStream = null;
        try {
            fileInputStream = new FileInputStream(file);
            byte[] readBytes = new byte[fileInputStream.available()];
            if (fileInputStream.read(readBytes) != -1) {
                String readString = new String(readBytes, Charset.forName(Constants.CHARSET));
                textView.setText(readString);
            }
        } catch (IOException e) {
            Log.e(TAG, e.toString());
        } finally {
            try {
                if (fileInputStream != null) {
                    fileInputStream.close();
                }
            } catch (IOException e) {
                Log.e(TAG, e.toString());
            }
        }
    }
}

File: UsbSerialConsole/app/src/main/java/jp/sugnakys/usbserialconsole/LogViewActivity.kt
package jp.sugnakys.usbserialconsole

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import jp.sugnakys.usbserialconsole.util.Constants
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.charset.Charset

class LogViewActivity : BaseAppCompatActivity() {
    private var logFile: File? = null
    private var toolbar: Toolbar? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.log_view_main)
        toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar
        if (actionBar != null) {
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            supportActionBar!!.setDisplayShowHomeEnabled(true)
            toolbar!!.setNavigationOnClickListener { finish() }
        }
    }

    override fun onResume() {
        super.onResume()
        logFile = intent.extras!![Constants.EXTRA_LOG_FILE] as File?
        if (logFile != null) {
            toolbar!!.title = logFile!!.name
            setLogText(logFile!!)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_log_view, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val intent: Intent
        when (item.itemId) {
            R.id.action_send_to -> {
                intent = Intent()
                intent.action = Intent.ACTION_VIEW
                intent.setDataAndType(Uri.fromFile(logFile), "text/plain")
                startActivity(intent)
            }
            else -> Log.e(TAG, "Unknown id")
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setLogText(file: File) {
        val textView = findViewById<View>(R.id.logView) as TextView
        var fileInputStream: FileInputStream? = null
        try {
            fileInputStream = FileInputStream(file)
            val readBytes = ByteArray(fileInputStream.available())
            if (fileInputStream.read(readBytes) != -1) {
                val readString = String(readBytes, Charset.forName(Constants.CHARSET))
                textView.text = readString
            }
        } catch (e: IOException) {
            Log.e(TAG, e.toString())
        } finally {
            try {
                fileInputStream?.close()
            } catch (e: IOException) {
                Log.e(TAG, e.toString())
            }
        }
    }

    companion object {
        private const val TAG = "LogViewActivity"
    }
}